### PR TITLE
Allow plugins to declare wildcard routes.

### PR DESCRIPTION
**Pull request by davidsoergel**
Currently, plugins must explicitly enumerate the routes that they serve via `get_plugin_apps()`.  This is inconvenient in the case of plugins that provide a set of related routes, such as a directory full of static files, whose names need not be known in advance.  This is the case, for instance, for dynamic plugins using Angular or React, where the build process produces a set of assets (including HTML, JS, CSS, source maps, etc.).

This PR allows a plugin to declare a wildcard route ending in `/*`, e.g. `{'/foo/*': self._foo_wildcard_handler}`.  Any request beginning with the specified path prefix will be passed to the given handler.

Note the handler may do whatever it likes with these requests; they do not necessarily correspond to filesystem paths (as in the static assets use case).  Thus this change also allows for RESTful apis, where object IDs and such are encoded as path elements.  Any security concerns (e.g., regarding paths containing `/../`), are the responsibility of the handler.

### CODE REVIEW

File: tensorboard/backend/application_test.py [hunk 1]
```diff
@@ -347,7 +347,7 @@ def _test(self, route, should_be_okay):
             plugin_name='foo',
             is_active_value=True,
             routes_mapping={route: lambda environ, start_response: None}),
-    ]
+      ]
```
Comment (wchargin) [inline tensorboard/backend/application_test.py]: Please revert (inconsistent indentation, and unrelated change).

File: tensorboard/backend/application_test.py [hunk 2]
```diff
@@ -535,6 +559,30 @@ def testNameToPluginMapping(self):
...
+    self._test_route('', False)
+
+  def testSlashlessRoute(self):
+    self._test_route('runaway', False)
+
+  def testGoodWildcardRoute(self):
+    self._test_route('/data/plugin/bar/wildcard/ok', True)
+  
+  def testBadWildcardRoute(self):
+    self._test_route('/data/plugin/bar/wildcard/bogus', False)
```
Comment (wchargin) [inline tensorboard/backend/application_test.py]: If I understand correctly, this test is a bit misleading: it suggests
that `…/wildcard/bogus` and `…/wildcard` are handled in the same way,
but they’re not—the former is handled by the plugin, while the latter
never resolves to a plugin app at all. There’s really nothing “bad”
about the `…/wildcard/bogus` route; the plugin handler just chooses to
reject it with the same status code.

What’s the purpose of having the `…/bogus` test case in addition to the
`…/ok` test case, given that they go through exactly the same codepaths
of the code under test?

File: tensorboard/backend/application.py [hunk 3]
```diff
@@ -328,7 +336,8 @@ def __call__(self, environ, start_response):  # pylint: disable=invalid-name
     """Central entry point for the TensorBoard application.
 
     This method handles routing to sub-applications. It does simple routing
-    using regular expression matching.
+    using strict string matching.  Regular expressions are not supported.
+    Wildcard routes such as `/foo/*` are supported as a special case.
```
Comment (wchargin) [inline tensorboard/backend/application.py]: As written, this implementation supports “`wildcard*`” routes even when
the wildcard is not its own path component, and so you could write a
wildcard route matching “`/foo1`”, “`/foo2`”, etc. Is this intentional,
or can we fail fast when a plugin declares a route like `/foo*`?

Separately, any thoughts as to disallowing asterisks in the middle of
routes, like `/foo/*/*` or `/bar/**/baz`, which will behave differently
than the author would expect?

These semantics should be documented on `TBPlugin.get_plugin_apps`,
whatever they end up being.

File: tensorboard/backend/application.py [hunk 4]
```diff
@@ -264,6 +264,14 @@ def __init__(self, plugins, path_prefix=''):
           path = (self._path_prefix + DATA_PREFIX + PLUGIN_PREFIX + '/' +
                   plugin.plugin_name + route)
         self.data_applications[path] = app
+    
+    # Wildcard routes will be checked in the given order, so we sort them
+    # longest to shortest so that a more specific route will take precedence
+    # over a more general one (e.g., a catchall route `/*` should come last).
+    self.prefix_routes = sorted(
+      [x for x in self.data_applications.keys() if x.endswith('*')],
```
Comment (wchargin) [inline tensorboard/backend/application.py]: This `.keys()` is superfluous.

File: tensorboard/backend/application.py [hunk 5]
```diff
@@ -264,6 +264,14 @@ def __init__(self, plugins, path_prefix=''):
           path = (self._path_prefix + DATA_PREFIX + PLUGIN_PREFIX + '/' +
                   plugin.plugin_name + route)
         self.data_applications[path] = app
+    
+    # Wildcard routes will be checked in the given order, so we sort them
+    # longest to shortest so that a more specific route will take precedence
+    # over a more general one (e.g., a catchall route `/*` should come last).
```
Comment (wchargin) [inline tensorboard/backend/application.py]: This seems important; want to test it?

File: tensorboard/backend/application_test.py [hunk 6]
```diff
@@ -506,20 +509,41 @@ def setUp(self):
               FakePlugin,
               plugin_name='bar',
               is_active_value=True,
-              routes_mapping={'/bar_route': self._bar_handler},
+              routes_mapping={
+                '/bar_route': self._bar_handler,
+                '/wildcard/*': self._wildcard_handler
+                },
```
Comment (wchargin) [inline tensorboard/backend/application_test.py]: nit: Please indent the closing delimiter of containers to the same level as
the line on which the container was opened:

```python
foo(
    routes_mapping={
        # ...
    },
    some_list=[
        # ...
    ],
)
```

File: tensorboard/backend/application_test.py [hunk 7]
```diff
@@ -506,20 +509,41 @@ def setUp(self):
...
 
   def _bar_handler(self):
     pass
 
-  def _construction_callback(self, context):
-    """Called when a plugin is constructed."""
-    self.context = context
+  @wrappers.Request.application
+  def _wildcard_handler(self, request):
+    if(request.path == '/data/plugin/bar/wildcard/ok'):
```
Comment (wchargin) [inline tensorboard/backend/application_test.py]: nit: Drop parentheses in Python.

File: tensorboard/backend/application.py (hunk 3 above)
Comment (nfelt) [reply tensorboard/backend/application.py]: +1 to forbidding asterisks anywhere else in the route, and to restricting them to follow a trailing `/` right now unless we have a good reason to allow generic wildcards.

Also, should we be creating a new `self.exact_routes` set for the non-wildcard routes that we can iterate over when we're doing the exact matching?  That way we won't match a request that comes in for a literal `/wildcard/*` path via the exact matching branch, which seems like a semantic error even if it's the same result in practice.

File: tensorboard/backend/application.py (hunk 5 above)
Comment (davidsoergel) [reply tensorboard/backend/application.py]: +1, added a test--thanks for the reminder!

File: tensorboard/backend/application.py (hunk 3 above)
Comment (davidsoergel) [reply tensorboard/backend/application.py]: Yep agreed throughout.  Refactored, replacing data_applications entirely with exact_routes and prefix_routes.  Added validation that there is at most one * (at the end only, and requiring a slash), and tests for that.

File: tensorboard/backend/application_test.py (hunk 1 above)
Comment (davidsoergel) [reply tensorboard/backend/application_test.py]: Done

File: tensorboard/backend/application_test.py (hunk 2 above)
Comment (davidsoergel) [reply tensorboard/backend/application_test.py]: This was really intended more as documentation than as a test per se (i.e., to show that the plugin may yet reject a request matching a wildcard).  Tried to clarify with naming and comments.

File: tensorboard/backend/application.py (hunk 4 above)
Comment (davidsoergel) [reply tensorboard/backend/application.py]: obsoleted.

File: tensorboard/backend/application_test.py (hunk 6 above)
Comment (davidsoergel) [reply tensorboard/backend/application_test.py]: done

File: tensorboard/backend/application_test.py (hunk 7 above)
Comment (davidsoergel) [reply tensorboard/backend/application_test.py]: done

File: tensorboard/backend/application.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (davidsoergel) [reply tensorboard/backend/application.py]: obsoleted

[the remark this answers is not in the document]
Comment (davidsoergel) [reply tensorboard/backend/application.py]: obsoleted

File: tensorboard/backend/application_test.py [hunk 8]
```diff
@@ -506,20 +509,41 @@ def setUp(self):
               FakePlugin,
               plugin_name='bar',
               is_active_value=True,
-              routes_mapping={'/bar_route': self._bar_handler},
+              routes_mapping={
+                '/bar_route': self._bar_handler,
+                '/wildcard/*': self._wildcard_handler
```
[the remark this answers is not in the document]
Comment (davidsoergel) [reply tensorboard/backend/application_test.py]: done

File: tensorboard/backend/application.py [hunk 9]
```diff
@@ -263,7 +264,25 @@ def __init__(self, plugins, path_prefix=''):
         else:
           path = (self._path_prefix + DATA_PREFIX + PLUGIN_PREFIX + '/' +
                   plugin.plugin_name + route)
-        self.data_applications[path] = app
+
+        if path.endswith('/*'):
+          # Note we remove the '*' but leave the slash in place.
+          path = path[:-1]
+          if '*' in path:
+            raise ValueError('Invalid route: only trailing wildcards are supported (i.e., `/.../*`)')
```
Comment (wchargin) [inline tensorboard/backend/application.py]: nit: Please wrap at 80:

```python
            raise ValueError(
                'Invalid route: only trailing wildcards are supported '
                '(i.e., `/.../*`)'
            )
```

Comment (wchargin) [inline tensorboard/backend/application.py]: Please include the text and plugin name in the error message. (Note that
in the `if`-case you’ll have to re-append the asterisk, or equivalent.)

File: tensorboard/backend/application.py [hunk 10]
```diff
@@ -345,9 +365,14 @@ class are WSGI applications.
     clean_path = _clean_path(parsed_url.path, self._path_prefix)
 
     # pylint: disable=too-many-function-args
-    if clean_path in self.data_applications:
-      return self.data_applications[clean_path](environ, start_response)
+    if clean_path in self.exact_routes:
+      return self.exact_routes[clean_path](environ, start_response)
     else:
+      for path_prefix in self.prefix_routes:
```
Comment (wchargin) [inline tensorboard/backend/application.py]: When I read this block, I look up to see how `self.prefix_routes` was
initialized: it’s initialized to `{}`, an unordered `dict`, so it’s not
clear that this `for`-loop’s iteration order is in fact well defined
because the attribute is later converted to an `OrderedDict`.

Maybe consider refactoring to populate a `raw_prefix_routes` local
variable, or at least add a comment at the original initialization of
`self.prefix_routes` to indicate that its type changes at runtime and
what the contract is?

File: tensorboard/backend/application.py [hunk 11]
```diff
@@ -345,9 +365,14 @@ class are WSGI applications.
...
+        if clean_path.startswith(path_prefix):
+          return self.prefix_routes[path_prefix](environ, start_response)
+        print(clean_path, ' does not start with ', path_prefix)
```
Comment (wchargin) [inline tensorboard/backend/application.py]: If we keep this around, it should be a logging statement, using a format
string so that it does not format the arguments when the message
wouldn’t be printed anyway.

File: tensorboard/backend/application_test.py [hunk 12]
```diff
@@ -535,6 +573,57 @@ def testNameToPluginMapping(self):
...
+    # A wildcard route requires a slash before the '*'.
+    # Lacking one, no route is matched.
+    self._test_route('/data/plugin/bar/wildcard', 404)
+
+  def testEmptyWildcardRouteWithSlash(self):
+    # A wildcard route requires a slash before the '*'.  Here we include the
+    # slash, so we might expect the route to match.
+    #
+    # However: Trailing slashes are automatically removed from incoming requests
+    # in _clean_path().  Consequently, this request does not match the wildcard
```
Comment (wchargin) [inline tensorboard/backend/application_test.py]: Hmm, interesting. Thanks for including this test case. Is this the
behavior that we actually want? It seems surprising that `/foo/*` does
not match `/foo/`: in glob contexts, `*` always matches the empty
string, so as a plugin author I would not expect this behavior.

The clean-path behavior of always identifying `/foo/` and `/foo` as
user-provided paths is certainly nice. I can see a case for making
pattern `/foo/*` match both paths `/foo/` and `/foo`, and then
disallowing plugins from having conflicting exact and wildcard routes
(e.g., separate `/foo/*` and `/foo/` handlers). How do you think that we
should address this?

File: tensorboard/backend/application_test.py (hunk 2 above)
Comment (wchargin) [reply tensorboard/backend/application_test.py]: Much clearer now; thanks!

File: tensorboard/backend/application_test.py (hunk 12 above)
Comment (nfelt) [reply tensorboard/backend/application_test.py]: I don't really like the behavior that clean_path has today, FWIW.  Conflating the request that has a trailing slash and the one without one in general is a bit dangerous because it encourages users to treat the URLs interchangeably, but the semantics of relative paths are different across the two URLs at which the content is served.

If we can establish it's safe, I'd be happy to remove that conversion logic or at least change it into an actual redirect so we aren't serving content under both URLs.  Basically, for every `/foo` handler we could create an implicit `/foo/` redirect handler.  We could either then block the plugin having both `/foo` and `/foo/` handlers (to be consistent), or potentially even block all handlers ending in `/`.

Comment (wchargin) [reply tensorboard/backend/application_test.py]: Oof, I was assuming that we already did have this redirect handler, à la
GitHub Pages, because in that case clean-path helps mitigate exactly the
confusion that you describe. Agreed, then.

File: tensorboard/backend/application.py (hunk 9 above)
Comment (davidsoergel) [reply tensorboard/backend/application.py]: Oops thanks.  I'm learning how dependent I've been on autoformatting... :P

Comment (davidsoergel) [reply tensorboard/backend/application.py]: done

File: tensorboard/backend/application.py (hunk 10 above)
Comment (davidsoergel) [reply tensorboard/backend/application.py]: done via unordered_prefix_routes.

File: tensorboard/backend/application.py (hunk 11 above)
Comment (davidsoergel) [reply tensorboard/backend/application.py]: Sorry, I just overlooked this debugging cruft.

File: tensorboard/backend/application_test.py (hunk 12 above)
Comment (davidsoergel) [reply tensorboard/backend/application_test.py]: Yep, makes sense, but can we punt fixing `/foo` vs. `/foo/` more generally to another PR?  If we leave this test in place, then it will fail when the behavior changes, so we won't forget to revisit it.

Comment (wchargin) [reply tensorboard/backend/application_test.py]: Okay—it’s not optimal, but if this change unblocks Angular/etc. then I’m
willing to defer.

File: tensorboard/backend/application.py [hunk 13]
```diff
@@ -265,7 +266,32 @@ def __init__(self, plugins, path_prefix=''):
...
+                             'trailing wildcards are supported '
+                             '(i.e., `/.../*`)' %
+                             (plugin.plugin_name, path))
+          self.exact_routes[path] = app
+
+    # Wildcard routes will be checked in the given order, so we sort them
+    # longest to shortest so that a more specific route will take precedence
+    # over a more general one (e.g., a catchall route `/*` should come last).
+    self.prefix_routes = collections.OrderedDict(
+        sorted(six.iteritems(unordered_prefix_routes),
```
Comment (wchargin) [inline tensorboard/backend/application.py]: nit: [Google Python style says][1] that when “using a hanging indent of
4 spaces […] there should be nothing after the open parenthesis or
bracket on the first line”.

[1]: https://google.github.io/styleguide/pyguide.html#34-indentation

```python
    self.prefix_routes = collections.OrderedDict(
        sorted(
            six.iteritems(unordered_prefix_routes),
            key=lambda x: len(x[0]),
            reverse=True))
```

File: tensorboard/backend/application_test.py [hunk 14]
```diff
@@ -543,27 +555,53 @@ def setUp(self):
...
+
+  def _bar_handler(self):
+    pass
+
+  @wrappers.Request.application
+  def _wildcard_handler(self, request):
+    if request.path == '/data/plugin/bar/wildcard/ok':
+      return wrappers.Response(response='hello world', status=200)
+    elif request.path == '/data/plugin/bar/wildcard/':
+      # this route cannot actually be hit; see testEmptyWildCardRouteWithSlash.
```
Comment (wchargin) [inline tensorboard/backend/application_test.py]: nit: caps (`testEmptyWildcardRouteWithSlash`)

File: tensorboard/backend/application.py [hunk 15]
```diff
@@ -265,7 +266,32 @@ def __init__(self, plugins, path_prefix=''):
...
           path = (self._path_prefix + DATA_PREFIX + PLUGIN_PREFIX + '/' +
                   plugin.plugin_name + route)
-        self.data_applications[path] = app
+
+        if path.endswith('/*'):
+          # Note we remove the '*' but leave the slash in place.
+          path = path[:-1]
+          if '*' in path:
+            # note we re-add the removed * in the format string
+            raise ValueError('Plugin %r handles invalid route %r*:  Only '
```
Comment (wchargin) [inline tensorboard/backend/application.py]: This will render as `invalid route '/foo/*/'*`, but the trailing
asterisk should be _inside_ the string.

Comment (wchargin) [inline tensorboard/backend/application.py]: nit: Parallel errors in this file (and generally throughout TensorBoard)
use one space after a period.

File: tensorboard/backend/application_test.py [hunk 16]
```diff
@@ -543,27 +555,53 @@ def setUp(self):
...
-    pass
+    self.server = werkzeug_test.Client(self.app, wrappers.BaseResponse)
 
   def _construction_callback(self, context):
     """Called when a plugin is constructed."""
     self.context = context
 
+  def _test_route(self, route, expected_status_code):
+    response = self.server.get(route)
+    self.assertEquals(response.status_code, expected_status_code)
```
Comment (wchargin) [inline tensorboard/backend/application_test.py]: The `assertEquals` method is deprecated; prefer `assertEqual`.

File: tensorboard/backend/application.py [hunk 17]
```diff
@@ -265,7 +266,32 @@ def __init__(self, plugins, path_prefix=''):
...
+                             'trailing wildcards are supported '
```
Comment (wchargin) [inline tensorboard/backend/application.py]: Nit, and I’m not requesting a change or blocking you on this, but please
note that some Google style guides explicitly discourage this kind of
horizontal alignment, because it means that changing the first line
forces you to reindent all the other lines for consistency. e.g.:
<https://google.github.io/styleguide/jsguide.html#formatting-horizontal-alignment>

Over time, as large-scale changes happen, necessarily either the blame
becomes polluted or the indentation rots; e.g.:

https://github.com/tensorflow/tensorboard/blob/757628a814410574718623b0a881ae10855754c8/tensorboard/plugins/pr_curve/pr_curve_demo.py#L45-L46

That file changed from `tf.flags` to `tf.compat.v1.flags` to `flags`,
and the alignment was only correct in the original form.

(If we adopt Black, per #2466, then this will be enforced. :-) )

File: tensorboard/backend/application.py (hunk 15 above)
Comment (davidsoergel) [reply tensorboard/backend/application.py]: done

Comment (davidsoergel) [reply tensorboard/backend/application.py]: done

File: tensorboard/backend/application.py (hunk 17 above)
Comment (davidsoergel) [reply tensorboard/backend/application.py]: Fair, but I just copied this from your https://github.com/tensorflow/tensorboard/blob/master/tensorboard/backend/application.py#L260 ;)

Anyway +100 to autoformatting.  I appreciate your rigor, but let's let automation handle this stuff :)  Many thanks for already doing the Prettier side of that!

File: tensorboard/backend/application.py (hunk 13 above)
Comment (davidsoergel) [reply tensorboard/backend/application.py]: done

File: tensorboard/backend/application_test.py (hunk 16 above)
Comment (davidsoergel) [reply tensorboard/backend/application_test.py]: done

File: tensorboard/backend/application_test.py (hunk 14 above)
Comment (davidsoergel) [reply tensorboard/backend/application_test.py]: done

File: tensorboard/backend/application_test.py (hunk 12 above)
Comment (davidsoergel) [reply tensorboard/backend/application_test.py]: OK.  I think it's topically different enough to merit a separate PR in any case.  Tracking at #2502.
